### PR TITLE
[WIP][Feature] introduce iceberg distributed plan

### PR DESCRIPTION
**Pull request by stephen-shelby**
Why I'm doing:

What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/connector/MetadataCollectJob.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import com.starrocks.thrift.TResultSinkType;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.velocity.VelocityContext;
+import org.apache.velocity.app.VelocityEngine;
+
+import java.io.StringWriter;
+
+// TODO(stephen): refactor and abstract metadata collect job
+public class MetadataCollectJob {
```
Comment (Smith-Cruise) [inline fe/fe-core/src/main/java/com/starrocks/connector/MetadataCollectJob.java:34]: If named `MetadataCollectJob` you should consider hudi/iceberg/xxx for future

File: be/src/runtime/metadata_result_writer.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        TIcebergMetadata iceberg_metadata;
+        meta_entries[i].__set_iceberg_metadata(iceberg_metadata);
+        auto& iceberg_meta = data_list[i].iceberg_metadata;
+
+        iceberg_meta.__set_content(content->get(i).get_int32());
+        iceberg_meta.__set_file_path(file_path->get_slice(i).to_string());
+        iceberg_meta.__set_file_format(file_format->get_slice(i).to_string());
+        iceberg_meta.__set_spec_id(spec_id->get(i).get_int32());
+
+        if (!columns[4]->is_null(i)) {
```
Comment (Smith-Cruise) [inline be/src/runtime/metadata_result_writer.cpp:161]: using constant name to replace number, easy for reviewing

File: fe/fe-core/src/main/java/com/starrocks/connector/MetadataCollectJob.java (hunk 1 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/connector/MetadataCollectJob.java:34]: yes, this is only wip patch. I will refactor it after review meeting

File: gensrc/thrift/Types.thrift [hunk 3]
```diff
@@ -397,7 +397,8 @@ enum TTableType {
     FILE_TABLE,
     DELTALAKE_TABLE,
     TABLE_FUNCTION_TABLE,
-    ODPS_TABLE
+    ODPS_TABLE,
+    LOGICAL_ICEBERG_METADATA_TABLE
```
Comment (DorianZheng) [inline gensrc/thrift/Types.thrift:401]: do we need to have different metadata table like `LOGICAL_HUDI_METADATA_TABLE` or `LOGICAL_DELTA_LAKE_METADATA_TABLE` in the future?

File: gensrc/thrift/PlanNodes.thrift [hunk 4]
```diff
@@ -1054,6 +1059,13 @@ struct THdfsScanNode {
     16: optional bool use_partition_column_value_only;
 
     17: optional Types.TTupleId mor_tuple_id;
+
+    // serialized static metadata table
+    18: optional string serialized_table;
+
+    // serialized origin table predicate for metaddata table scan filter
+    19: optional string serialized_predicate;
```
Comment (DorianZheng) [inline gensrc/thrift/PlanNodes.thrift:1067]: why do we need this when we already have `conjuncts`?

**Comment by DorianZheng** [comment]
Gread job!

File: fe/fe-core/src/main/java/com/starrocks/planner/IcebergMetadataScanNode.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+        long snapshotId = -1;
+        // TODO(stephen): parse version by AstBuilder
+        if (!Strings.isNullOrEmpty(temporalClause)) {
+            String kw = "version as of";
+            temporalClause = temporalClause.substring(kw.length()).trim();
+            snapshotId = Long.parseLong(temporalClause);
+        }
+
+        SerializedMetaSpec SerializedMetaSpec = GlobalStateMgr.getCurrentState().getMetadataMgr()
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/planner/IcebergMetadataScanNode.java:82]: do you think it's better to reuse ` GlobalStateMgr.getCurrentState().getMetadataMgr().getRemoteFileInfos` which means that let's treat meatadata table as a connector

**Pull request closed without merging**